ntoskrnl/io/pnpmgr/pnpnotify.c

```diff
@@ -19,210 +19,259 @@
-                TAG_PNP_NOTIFY);
-            if (!NotificationInfos)
+            PDEVICE_INTERFACE_CHANGE_NOTIFICATION notifStruct = NotificationStructure;
+
+            if (!IsEqualGUID(&notifStruct->InterfaceClassGuid, &Entry->Guid))
             {
-                KeReleaseGuardedMutex(&PnpNotifyListLock);
                 return;
             }
-            NotificationInfos->Version = 1;
-            NotificationInfos->Size = sizeof(HWPROFILE_CHANGE_NOTIFICATION);
-            RtlCopyMemory(&NotificationInfos->Event, Event, sizeof(GUID));
             break;
         }
         case EventCategoryTargetDeviceChange:
         {
-            if (Event != &GUID_PNP_CUSTOM_NOTIFICATION)
-            {
-                PTARGET_DEVICE_REMOVAL_NOTIFICATION NotificationInfos;
-                NotificationStructure = NotificationInfos = ExAllocatePoolWithTag(
-                    PagedPool,
-                    sizeof(TARGET_DEVICE_REMOVAL_NOTIFICATION),
-                    TAG_PNP_NOTIFY);
-                if (!NotificationInfos)
-                {
-                    KeReleaseGuardedMutex(&PnpNotifyListLock);
-                    return;
-                }
-                NotificationInfos->Version = 1;
-                NotificationInfos->Size = sizeof(TARGET_DEVICE_REMOVAL_NOTIFICATION);
-                RtlCopyMemory(&NotificationInfos->Event, Event, sizeof(GUID));
-            }
-            else
+            PTARGET_DEVICE_REMOVAL_NOTIFICATION notifStruct = NotificationStructure;
+
+            if (DeviceObject != Entry->FileObjectPDO)
             {
-                PTARGET_DEVICE_CUSTOM_NOTIFICATION NotificationInfos;
-                NotificationStructure = NotificationInfos = ExAllocatePoolWithTag(
-                    PagedPool,
-                    sizeof(TARGET_DEVICE_CUSTOM_NOTIFICATION),
-                    TAG_PNP_NOTIFY);
-                if (!NotificationInfos)
-                {
-                    KeReleaseGuardedMutex(&PnpNotifyListLock);
-                    return;
-                }
-                RtlCopyMemory(NotificationInfos, EventCategoryData1, sizeof(TARGET_DEVICE_CUSTOM_NOTIFICATION));
+                return;
             }
+            notifStruct->FileObject = Entry->FileObject;
             break;
         }
         default:
         {
-            DPRINT1("IopNotifyPlugPlayNotification(): unknown EventCategory 0x%x UNIMPLEMENTED\n", EventCategory);
-            KeReleaseGuardedMutex(&PnpNotifyListLock);
-            return;
+            break;
         }
     }
 
-    /* Loop through procedures registred in PnpNotifyListHead
-     * list to find those that meet some criteria.
-     */
-    ListEntry = PnpNotifyListHead.Flink;
-    while (ListEntry != &PnpNotifyListHead)
+    /* Call entry into new allocated memory */
+    DPRINT("IopNotifyPlugPlayNotification(): found suitable callback %p\n", Entry);
+
+    KeReleaseGuardedMutex(&PiNotificationInProgressLock);
```

Comment:
Why is this lock released here? Or better: Why is this function called with this lock held? And why does the name imply that it's held while the notification is in progress, rather than protecting the list, which afaics is the only thing it does.

Reply:
The lock is released because we need to call the routine with APCs enabled. It's not stated explicitly in the docs but it's "what windows does" and in this case such behaviour should be kept, I think.
I'll add a comment if we keep the lock in the end

Reply:
Looks like I've found a problem here. The element can be deleted while we're inside `PnpNotificationProc` (by another thread) because the lock is not held.